cads_adaptors/adaptors/cds.py

```diff
@@ -115,38 +85,143 @@ def estimate_costs(
+        """
+        Normalise the request prior to submission to the broker, and at the start of the retrieval.
+        This is executed on the retrieve-api pod, and then repeated on the worker pod.
+
+        The returned request needs to be compatible with the web-portal, it is currently what is used
+        on the "Your requests" page, hence it should not be modified to much from the user's request.
+        """
+        if self.normalised:
+            return request
+        
+        # Make a copy of the original request for debugging purposes
+        self.input_request = deepcopy(request)
+        self.context.debug(f"Input request:\n{self.input_request}")
+
+        # Apply any pre-mapping modifications
+        working_request = self.pre_mapping_modifications(deepcopy(request))
+
+        # If specified by the adaptor, intersect the request with the constraints.
+        # The intersected_request is a list of requests
+        if self.intersect_constraints_bool:
+            self.request_intersected = self.intersect_constraints(working_request)
+            if len(self.request_intersected) == 0:
+                msg = "Error: no intersection with the constraints."
+                self.context.add_user_visible_error(message=msg)
+                raise InvalidRequest(msg)
+        else:
+            self.request_intersected = ensure_list(working_request)
+
+        # Map the list of requests
+        self.mapped_requests = [
+            self.apply_mapping(i_request) for i_request in self.request_intersected
+        ]
+
+        # Implement embargo if specified
+        if self.embargo is not None:
+            from cads_adaptors.tools.date_tools import implement_embargo
+            try:
+                self.mapped_requests, cacheable_embargo = implement_embargo(
+                    self.mapped_requests, self.embargo
+                )
+            except ValueError as e:
+                self.context.add_user_visible_error(
+                    message=f"{e}"
+                )
+                raise InvalidRequest(e)
+
+            if not cacheable_embargo:
+                # Add an uncacheable key to the request
+                random_key = str(randint(0, 2**128))
+                request["_part_of_request_under_embargo"] = random_key
+
+
         schemas = self.schemas
         if not isinstance(schemas, list):
             schemas = [schemas]
         # Apply first dataset schemas, then adaptor schema
         if adaptor_schema := self.adaptor_schema:
             schemas = schemas + [adaptor_schema]
         for schema in schemas:
-            request = enforce.enforce(request, schema, self.context.logger)
-        if not isinstance(request, dict):
-            raise TypeError(
-                f"Normalised request is not a dictionary, instead it is of type {type(request)}"
-            )
+            self.mapped_requests = [
+                enforce.enforce(i_request, schema, self.context.logger)
```

Comment:
As far as I know, at least some (if not all) schemas need to be enforced on the initial request. Maybe they could be enforced on the requests resulted after intersecting the constraints (even though that would likely just repeat the same effort that could be spent only on the initial request). In any case, applying them on the already mapped request(s) seems a bit too late in the chain of transformations. This is the case with at least one CAMS adaptor (https://github.com/ecmwf-projects/cads-adaptors/blob/main/cads_adaptors/adaptors/cams_solar_rad/__init__.py#L19).

Reply:
It has been concluded that both schemas on the input request and the mapped requests make sense. For now only the first type is supported.